packages/socket-mode/src/SocketModeClient.ts

```diff
@@ -402,27 +116,238 @@ export class SocketModeClient extends EventEmitter {
+   * unless you call start() again later.
    */
   public disconnect(): Promise<void> {
     return new Promise((resolve, reject) => {
-      this.logger.debug('manual disconnect');
-
-      // resolve (or reject) on disconnect
+      this.logger.debug('Manually disconnecting this Socket Mode client');
+      // Resolve (or reject) on disconnect
       this.once('disconnected', (err) => {
         if (err instanceof Error) {
           reject(err);
         } else {
           resolve();
         }
       });
-
-      // delegate behavior to state machine
+      // Delegate behavior to state machine
       this.stateMachine.handle('explicit disconnect');
     });
   }
 
+  // --------------------------------------------
+  // Private methods / properties
+  // --------------------------------------------
+
+  /**
+   * State machine that backs the transition and action behavior
+   */
+  private stateMachine: StateMachine<string, string>;
+
+  /* eslint-disable @typescript-eslint/indent, newline-per-chained-call */
+
+  private connectingSubmachine: Configuration<string, string> = Finity.configure()
+    .global()
+      .onStateEnter((state) => {
+        this.logger.debug(`Transitioning to state: connecting:${state}`);
+      })
+    .initialState('authenticating')
+      .do(this.retrieveWSSURL.bind(this))
+        .onSuccess().transitionTo('authenticated')
+        .onFailure()
+          .transitionTo('reconnecting').withCondition(this.reconnectingCondition.bind(this))
+          .transitionTo('failed')
+    .state('authenticated')
+      .onEnter(this.configureAuthenticatedWebSocket.bind(this))
+      .on('websocket open').transitionTo('handshaking')
+    .state('handshaking') // a state in which to wait until the 'server hello' event
+    .state('failed')
+      .onEnter(this.handleConnectionFailure.bind(this))
+  .getConfig();
+
+  private connectedSubmachine: Configuration<string, string> = Finity.configure()
+    .global()
+      .onStateEnter((state) => {
+        this.logger.debug(`Transitioning to state: connected:${state}`);
+      })
+    .initialState('ready')
+      .onEnter(() => {
+        if (this.badConnection) {
+          // arrived here because `server ping timeout` occurred and a new connection was created
+          // tear down old connection
+          this.tearDownWebSocket();
+          this.badConnection = false;
+        }
+        // Start heartbeat to keep track of the websocket connection continuing to be alive
+        this.startPeriodicallySendingPingToSlack();
+        this.startMonitoringPingFromSlack();
+        // The transition isn't done yet, so we delay the following statement until after the event loop returns
+        setImmediate(() => {
+          this.emit('ready');
+        });
+      })
+      .on('server disconnect warning')
+        .transitionTo('connecting').withCondition(() => this.autoReconnectEnabled)
```

Comment:
Changed from `refreshing-connection` to `connecting`. The whole logic is much simpler now.